### ISSUE TITLE
Request: Lint rule to enforce ordering of docstring sections

### ISSUE THREAD

**Issue by stinodego** [astral-sh/ruff#9641]
Our project follows the NumPy docstring conventions outlined [here](https://numpydoc.readthedocs.io/en/latest/format.html).

One extremely common mistake our contributors make is the order in which docstring sections appear. Often I'll have to correct docstrings where people put the `Returns` section after the `Examples` section, or the `Notes` section before the `Warnings` section.

Could ruff autodetect/autofix this?

**Before**
```python
def cool_function() -> int:
    """
    Description.

    Notes
    -----
    Important edge case.

    Warnings
    --------
    Be aware of this requirement!

    Returns
    -------
    int
    """
```

**After**
```python
def cool_function() -> int:
    """
    Description.

    Returns
    -------
    int

    Warnings
    --------
    Be aware of this requirement!

    Notes
    -----
    Important edge case.
    """
```

**Comment by charliermarsh** [comment]
Yeah, we should definitely lint for this given that it's part of the docstring spec.

Since pydocstyle has been deprecated, I would be comfortable adding this to the `D` rules directly. We could enable it for the NumPy convention (but disable it for the Google and PEP 257 conventions).

**Comment by dpgraham4401** [comment]
@charliermarsh, to clarify, you're thinking this should be added this as `D420` since `D419` is the last error code in the [pydocstyle Docstring Content Issues](https://www.pydocstyle.org/en/stable/error_codes.html#grouping)? grouping

**Comment by dhruvmanila** [comment]
> to clarify, you're thinking this should be added this as `D420` since `D419` is the last error code in the [pydocstyle Docstring Content Issues](https://www.pydocstyle.org/en/stable/error_codes.html#grouping)?

I'd say yes as `D4XX` are for docstring content issues (https://www.pydocstyle.org/en/stable/error_codes.html#grouping).

**Comment by sebastian-altamirano** [comment]
The rule should also work for Google-style docstrings:
- https://google.github.io/styleguide/pyguide.html#383-functions-and-methods
- https://github.com/google/styleguide/issues/847#issuecomment-2513517520

**Comment by dpgraham4401** [comment]
I'm [working on this](https://github.com/dpgraham4401/ruff/blob/numpy_docstring_ordering-9641/crates/ruff_linter/src/rules/pydocstyle/rules/docstring_order.rs), but it'll be my first time contributing to ruff and I'm a novice rustacean so might be a while before a PR 😁

**Comment by dpgraham4401** [comment]
Update, life has gotten in the way and I'm not going to have time to work on this for some time. Sorry y'all.

**Comment by o1x3** [comment]
Hey, couple design decisions I want to get right before opening this up.

1. How strict for Google convention?

The Google style guide only actually specifies Args before Returns/Yields, and Raises after those. A Google maintainer confirmed this ([link]). So I'm leaning toward only enforcing those three constraints and leaving everything else (Notes, Examples, Attributes, etc.) unordered. Nobody gets a false positive because they put Examples before Notes or whatever.

The other option is enforcing a full ordering across all ~14 section groups based on Napoleon/Sphinx conventions, but that feels more opinionated than the style guide warrants. Happy to be talked out of this.

2. Auto-detection behavior?

When `convention` isn't set, ruff auto-detects docstring style for D405-D417. Should D420 do the same, or only fire when someone explicitly sets `convention = "google"` / `"numpy"`? The issue thread mentioned enabling for NumPy and later Google but didn't really cover the unset case. No strong opinion here.


---

For reference, the current implementation:
- Preview rule (`preview_since = "NEXT_RUFF_VERSION"`)
- No autofix (diagnostic only)
- NumPy ordering follows the [numpydoc spec](https://numpydoc.readthedocs.io/en/latest/format.html) exactly
- PEP257 convention ignores D420

**Comment by ntBre** [comment]
That all sounds reasonable to me. I think I would only enforce the Google styles that are explicitly documented, and it seems reasonable to auto-detect the convention like in the other rules.

**Comment by stinodego** [comment]
@o1x3 Thanks a lot for implementing this! Looking forward to using it in the near future 😊
